dlpack: Do not call PyBuffer_Release on the exporter object

dalcinl opened the pull request:
Fixes #609 

cc @snsunx-nv @seberg

seberg commented on src/mpi4py/MPI.src/asbuffer.pxi:
```diff
@@ -75,6 +76,10 @@ cdef int PyMPI_GetBuffer(object obj, Py_buffer *view, int flags) except -1:
 
+cdef void PyMPI_ReleaseBuffer(int kind, Py_buffer *view) noexcept:
+    if kind != 0: Py_CLEAR(view.obj)
+    PyBuffer_Release(view)
```
I might suggest to make this an if/else clause.  `PyBuffer_Release` shouldn't do anything additional here and that way it is more obvious that this is custom use of `Py_buffer` that comes with custom cleanup.

But the change looks good to me to fix the cupy issue.

leofang commented:
Commit e6a8273 looks fine to me, though I imagine at some point we want to hold the capsule as the exporting object instead. This PR includes other commits which does not seem relevant to the fix, I assume this is for preparing a v4.0.3 release?

dalcinl commented:
> I imagine at some point we want to hold the capsule as the exporting object instead.

Yes, sure. This is a quick fix for a patch release. I'm working on the improvements, but it turns out that my ctypes-based DLPack testing is causing me a huge headache. ctypes messes with Python object references in a rather opaque way.

> This PR includes other commits which does not seem relevant to the fix,

I simply screwed up and rebased to master. Fixed now.

leofang commented:
Thanks, Lisandro and all!

leofang commented:
> but it turns out that my ctypes-based DLPack testing is causing me a huge headache. ctypes messes with Python object references in a rather opaque way.

leofang commented:
> but it turns out that my ctypes-based DLPack testing is causing me a huge headache. ctypes messes with Python object references in a rather opaque way.

I see, so the stuff we discussed is still not fixed...?

seberg commented:
You can ping me if you have some code with ref-counting bugs on this specific issue.  I work with the Python C-API so much, that I might have a better eye to just spot whats off.

dalcinl commented:
@seberg @leofang My issues are actually with our pure-Python ctypes-based code used to test the DLPack support, without having to depend on third-party producer library. Leo tried to reuse that code long ago, but he failed, and now I realize that it was probably working on mpi4py's side just by accident.

The root of the issues is in ctypes. If you set a capsule destructor callback implemented in Python with ctypes, then there are two details that make it tricky to work: the destructor is called when the capsule reference count reached zero, but then the Python runtime calls the destructor which happens to be implemented in Python, therefore the capsule should be (temporarily) resurrected with incref and eventually decrefed again, which leads to a recursive call you have to cope with. But before dealing with that problem, there is an even more pressing second issue: some ctypes calls (eg. `ctypes.cast`) do not seem to work well if the error indicator (`PyErr_Occured() != NULL`) is set, and this totally defeats my testing purposes, which includes the proper cleanup of Python references in the case of exceptions.

I managed to workaround all these problems by just not attaching a desctructor to the capsules I create. IOW, the capsules just borrow a reference to `DLManagedTensor`. This means that my ctypes code is not really general as a dlpack implementation, but it is good enough to test mpi4py and ensure 100% Cython code coverage.

PS: I'll ping you guys back shortly to review a new PR where I fix things to hold on the capsule rather than the original producer.
